File: ANN/matlab_isachelp/matlab/assignment3/problem3/main.m

TGamesTrain = 4*10^3;
mu = 0.8;
mucopy = mu;
a = 0.9;
Q1=[];
Q2=[];

winnerList=zeros(TGamesTrain,1);
for iGame = 1:TGamesTrain
    if mod(iGame,1000)==0
       fprintf('game: %0.0f',iGame) 
    end
    board = zeros(3,3);
    board2 = board;
    
    for iRound=1:5
        [bool,i] = doesStateExist(Q1,board);
        if bool
            q = Q1(4:6,3*i-2:3*i); 
        else
            Q1(1:3,end+1:end+3) = board;
            q = NaN(3);
            q(1,board(1,:)==0) = 0; 
            q(2,board(2,:)==0) = 0; 
            q(3,board(3,:)==0) = 0; 
            Q1(4:6,end-2:end) = q;
        end
        [i1,i2] = newPos(q,mu);
        board(i1,i2) = 1;
        board2(i1,i2) = 2*iRound-1;
        if hasWon(board) break, end
        if iRound == 5 break, end
        
        %player2
        [bool,i] = doesStateExist(Q2,board);
        if bool
            q = Q2(4:6,3*i-2:3*i);
        else
            Q2(1:3,end+1:end+3) = board;
            q = NaN(3);
            q(1,board(1,:)==0) = 0; %random values betwen 0-0.1!
            q(2,board(2,:)==0) = 0; 
            q(3,board(3,:)==0) = 0; 
            Q2(4:6,end-2:end) = q;
        end
        [i1,i2] = newPos(q,mu);
        board(i1,i2) = -1;
        board2(i1,i2) = 2*iRound;
        
        if hasWon(board) break, end
    end    
    
    %update qMatrix
    maxIndex = max(max(board2));
    winner = mod(maxIndex,2)*2-1;
    if maxIndex == 9
        if not(hasWon(board))
            winner=0;
        end
    end
    winnerList(iGame)=winner;
    for i=1:maxIndex-1
       b1 = (board2<=i-1).*board2;
       b2 = (board2<=i+1).*board2;
       previousBoard = mod(b1,2)*2-1;
       nextBoard =  mod(b2,2)*2-1;
       if mod(i,2)==0
           [~,i1] = doesStateExist(Q1,previousBoard);
           [~,i2] = doesStateExist(Q1,nextBoard);
           q1 = Q1(4:6,3*i1-2:3*i1);
           q2 = Q1(4:6,3*i2-2:3*i2);
           mQ = max(max(q2));
           maxNextState = (q2==mQ)*mQ;
           newQ = q1+a*(maxNextState-q1);
           Q1(4:6,3*i1-2:3*i1) = newQ;
       else
           [~,i1] = doesStateExist(Q2,previousBoard);
           [~,i2] = doesStateExist(Q2,nextBoard);
           q1 = Q2(4:6,3*i1-2:3*i1);
           q2 = Q2(4:6,3*i2-2:3*i2);
           mQ = max(max(q2));
           maxNextState = (q2==mQ)*mQ;
           newQ = q1+a*(maxNextState-q1);
           Q2(4:6,3*i1-2:3*i1) = newQ;
       end
    end
    
    if winner == 1 || winner ==0
       board(board2==maxIndex)=0;
       [~,i1] = doesStateExist(Q1,board);
       Q1(4:6,i1*3-2:i1*3) = Q1(4:6,i1*3-2:i1*3)+winner*a*(board2==maxIndex);
       board(board2==maxIndex-1)=0;
       [~,i1] = doesStateExist(Q2,board);
       Q2(4:6,i1*3-2:i1*3) = Q2(4:6,i1*3-2:i1*3)+winner*a*(board2==(maxIndex-1));   
    else
       board(board2==maxIndex)=0;
       [~,i1] = doesStateExist(Q2,board);
       Q2(4:6,i1*3-2:i1*3) = Q2(4:6,i1*3-2:i1*3)+winner*a*(board2==maxIndex);
       board(board2==maxIndex-1)=0;
       [~,i1] = doesStateExist(Q1,board);
       Q1(4:6,i1*3-2:i1*3) = Q1(4:6,i1*3-2:i1*3)+winner*a*(board2==(maxIndex-1));  
    end

    mu = mu-mucopy/TGamesTrain;
end
plot(winnerList)
csvwrite('player1.csv',Q1);
csvwrite('player2.csv',Q1);

File: ANN/matlab_isachelp/matlab/assignment3/problem3/python/main.py
from Qtable import Qt 
import numpy as np
import time
from matplotlib import pyplot as plt
import csv
import random


def PlayerhasWon(board):
    for x in range(3):
        if 3==abs(sum(board[x,:])):
            return True
    for y in range(3):
        if 3==abs(sum(board[:,y])):
            return True
    c1 = 3==abs(board[0,0]+board[1,1]+board[2,2])
    c2 = 3==abs(board[2,0]+board[1,1]+board[0,2])
    return c1 or c2

TGamesTrain = 1*pow(10,4)
TGamesTest = pow(10,3)
mu = 0.3
mucopy = 0.3
alpha=0.3
Q1=Qt(1,alpha)
Q2=Qt(-1,alpha)

for iGame in range(TGamesTrain):
    mu-=mucopy/TGamesTrain
    board = np.array([[0,0,0],[0,0,0],[0,0,0]])
    boardOrder= np.array([[0,0,0],[0,0,0],[0,0,0]])
    for iRound in range(5):
        board,boardOrder = Q1.UpdateBoards(board,boardOrder,mu)
        if PlayerhasWon(board):hasWon=1; break
        if iRound==4:hasWon=0; break
        board,boardOrder = Q2.UpdateBoards(board,boardOrder,mu)
        if PlayerhasWon(board):hasWon=1; break
    if hasWon:
        Q1.UpdateQvalues(board,boardOrder,1,False)
        Q2.UpdateQvalues(board,boardOrder,2,False)
    else:
        Q1.UpdateQvalues(board,boardOrder,1,True)
        Q2.UpdateQvalues(board,boardOrder,2,True)
    if iGame%1000==0: print(iGame)

print('Training Complete')
victories=[]
for iGame in range(TGamesTest):
    board = np.array([[0,0,0],[0,0,0],[0,0,0]])
    boardOrder= np.array([[0,0,0],[0,0,0],[0,0,0]])
    for iRound in range(5):
        board,boardOrder = Q1.UpdateBoards(board,boardOrder,mu)
        if PlayerhasWon(board):hasWon=1; break
        if iRound==4:hasWon=0; break
        board,boardOrder = Q2.UpdateBoards(board,boardOrder,mu)
        if PlayerhasWon(board):hasWon=1; break
    if hasWon:
        player = (np.max(boardOrder)%2)*2-1
        victories.append(player)
    else:
        victories.append(0)
print('Test Complete')
plt.plot(victories)
plt.show() 



f = open('Q1.csv', 'w')
writer = csv.writer(f)
for i in range(6):
    row=Q1.Qtable[i,:]
    c = [np.nan if x==Q1.nan else x for x in row]
    writer.writerow(c)
f.close()

f = open('Q2.csv', 'w')
writer = csv.writer(f)
for i in range(6):
    row=Q2.Qtable[i,:]
    c = [np.nan if x==Q2.nan else x for x in row]
    writer.writerow(c)
f.close()

#play 3 games to check:
for iGame in range(3):
    board = np.array([[0,0,0],[0,0,0],[0,0,0]])
    boardOrder= np.array([[0,0,0],[0,0,0],[0,0,0]])
    for iRound in range(5):
        print(f'board: {board}')
        inputPlayer = int(input())-1
        y=inputPlayer//3
        x=inputPlayer%3
        board[y,x]=1
        boardOrder[y,x]=np.max(boardOrder)+1
        if PlayerhasWon(board):hasWon=1; break
        if iRound==4:hasWon=0; break
        board,boardOrder = Q2.UpdateBoards(board,boardOrder,mu)
        if PlayerhasWon(board):hasWon=1; break
    if hasWon:
        print('someone Won')
    else:
        print('draw')

